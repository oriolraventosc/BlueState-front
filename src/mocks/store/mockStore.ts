import { configureStore } from "@reduxjs/toolkit";
import { contactsReducer } from "../../redux/features/BusinessSlice/BusinessSlice";
import { toastSliceReducer } from "../../redux/features/UiSlice/UiSlice";
import {
  userReducer,
  UserState,
} from "../../redux/features/UserSlice/UserSlice";

interface MockStoreProps {
  userPreloadedState: UserState;
}

const mockStore = ({ userPreloadedState }: MockStoreProps) =>
  configureStore({
    reducer: {
      userActions: userReducer,
      uiModal: toastSliceReducer,
      contactsActions: contactsReducer,
    },
    preloadedState: {
      userActions: userPreloadedState,
    },
  });

export default mockStore;
