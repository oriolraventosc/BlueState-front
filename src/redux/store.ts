import { configureStore, ThunkAction, Action } from "@reduxjs/toolkit";
import { userReducer } from "./features/UserSlice/UserSlice";
import { toastSliceReducer } from "./features/UiSlice/UiSlice";
import { contactsReducer } from "./features/BusinessSlice/BusinessSlice";

export const store = configureStore({
  reducer: {
    userActions: userReducer,
    uiModal: toastSliceReducer,
    contactsActions: contactsReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
