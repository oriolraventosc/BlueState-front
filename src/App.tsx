import React, { Suspense } from "react";
import "./App.css";
import Layout from "./components/Layout/Layout";
import Loading from "./components/Loading/Loading";
import Modal from "./components/Modal/Modal";
import useToken from "./hooks/useToken/useToken";
import { useAppSelector } from "./redux/hooks";

const App = () => {
  const { getToken } = useToken();
  const alert = useAppSelector(
    ({ uiModal: uiSuccessModal }) => uiSuccessModal.alert
  );
  const isLoading = useAppSelector(({ uiModal }) => uiModal.isLoading);

  getToken();
  return (
    <>
      {alert && <Modal />}
      <Suspense fallback={<Loading />} />
      {isLoading && <Loading />}
      <div className="app">
        <Layout />
      </div>
    </>
  );
};

export default App;
