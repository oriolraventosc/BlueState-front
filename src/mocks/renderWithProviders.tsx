import { configureStore, PreloadedState } from "@reduxjs/toolkit";
import { render, RenderOptions } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { PropsWithChildren } from "react";
import { Provider } from "react-redux";
import { toastSliceReducer } from "../redux/features/UiSlice/UiSlice";
import { RootState, store } from "../redux/store";
import { userReducer } from "../redux/features/UserSlice/UserSlice";
import { InitialEntry } from "@remix-run/router";
import { contactsReducer } from "../redux/features/BusinessSlice/BusinessSlice";

interface ExtendedRenderOptions extends RenderOptions {
  preloadedState?: PreloadedState<RootState>;
  store?: typeof store;
  initialEntries?: InitialEntry[];
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    preloadedState,
    initialEntries,
    store = configureStore({
      reducer: {
        uiModal: toastSliceReducer,
        userActions: userReducer,
        contactsActions: contactsReducer,
      },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  const Wrapper = ({ children }: PropsWithChildren<{}>): JSX.Element => {
    return (
      <>
        <BrowserRouter>
          <Provider store={store}>{children}</Provider>
        </BrowserRouter>
      </>
    );
  };
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};

export default renderWithProviders;
