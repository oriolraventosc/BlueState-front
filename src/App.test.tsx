import App from "./App";
import TestRenderer from "react-test-renderer";
import { Provider } from "react-redux";
import { store } from "./redux/store";
import { MemoryRouter } from "react-router-dom";
import mockStore from "./mocks/store/mockStore";
import {
  mockUserLogged,
  mockUserNotLogged,
} from "./mocks/states/mockUserState";
import renderWithProviders from "./mocks/renderWithProviders";
import { screen, waitFor } from "@testing-library/react";

describe("Given the App component", () => {
  describe("When it is rendered with a path '/*'", () => {
    test("Then it should show the login page", () => {
      const expected = TestRenderer.create(
        <Provider store={store}>
          <MemoryRouter initialEntries={["/"]}>
            <App />
          </MemoryRouter>
        </Provider>
      );

      expect(expected).toMatchSnapshot();
    });
  });

  describe("When it is rendered and the user is logged", () => {
    test("Then it should show the contacts page", async () => {
      const initialEntries = ["/contacts"];
      const store = mockStore({ userPreloadedState: mockUserLogged });

      renderWithProviders(<App />, { store, initialEntries });
      await waitFor(() => {
        const text = screen.getByText("Made by");

        expect(text).toBeInTheDocument();
      });
    });
  });

  describe("When it is rendered with the path '/contacts' and the user is not logged", () => {
    test("Then it should show the login page", async () => {
      const initialEntries = ["/contacts"];
      const store = mockStore({ userPreloadedState: mockUserNotLogged });

      renderWithProviders(<App />, { store, initialEntries });

      await waitFor(() => {
        const title = screen.queryByRole("heading", {
          name: "Sign in",
        });

        expect(title).toBeInTheDocument();
      });
    });
  });

  describe("When it is rendered with the path '/login' and the user is logged", () => {
    test("Then it should show the contacts page", async () => {
      const initialEntries = ["/login"];
      const store = mockStore({ userPreloadedState: mockUserLogged });

      renderWithProviders(<App />, { store, initialEntries });
      await waitFor(() => {
        const text = screen.getByText("Made by");

        expect(text).toBeInTheDocument();
      });
    });
  });

  describe("When it is rendered with the path '/register' and the user is logged", () => {
    test("Then it should show the contacts page", async () => {
      const initialEntries = ["/register"];
      const store = mockStore({ userPreloadedState: mockUserLogged });

      renderWithProviders(<App />, { store, initialEntries });
      await waitFor(() => {
        const text = screen.getByText("Made by");

        expect(text).toBeInTheDocument();
      });
    });
  });
});
