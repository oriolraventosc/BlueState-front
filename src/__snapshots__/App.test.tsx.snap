// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the App component When it is rendered with a path '/*' Then it should show the login page 1`] = `
<div
  className="app"
/>
`;
